Remove v2 schema1 push

tiborvass opened the pull request:
Manifest v2 schema1 was deprecated in 4866f5139a0b0bc026aa4b0c916ee1071c695134 (#39365) and this PR
removes the push code for v2 schema1.

cc @cpuguy83 @dmcgowan @tonistiigi @thaJeztah @andrewhsu

Commit:
Revert "Revert "Remove the rest of v1 manifest support""

This reverts commit f695e98cb7cbb39eb7e5bd0a268ae1cd2e594fc1.

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
Revert "use gotest.tools assertions in docker_cli_push_test.go"

This reverts commit 0811297608499e8be473473b80eaac9d2ced9dd8.

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
distribution: remove v2 schema1 push

Manifest v2 schema1 was deprecated in 4866f5139a1 and this commit
removes the push code for v2 schema1.

Signed-off-by: Tibor Vass <tibor@docker.com>

Commit:
daemon: do not mkdir trust directory

Signed-off-by: Tibor Vass <tibor@docker.com>

thaJeztah reviewed (commented):
Should we copy the functions from libtrust, and remove the dependency?

tiborvass commented:
@thaJeztah I tried to do that, but distribution code also depends on libtrust. I believe the dependency can be completely removed with containerd integration.

thaJeztah commented on integration/system/uuid_test.go:
```diff
@@ -0,0 +1,23 @@
+	info, err := c.Info(context.Background())
+	assert.NilError(t, err)
+
+	_, err = uuid.Parse(info.ID)
```
Wondering to what extent we need to check the actual format, given https://github.com/moby/moby/blob/18b23067be82054a6384fc5934ecbb0276d1862b/api/swagger.yaml#L3365-L3370

Orthogonal to this; we should look at what purpose this field now serves. In a multi-node setup (Swarm cluster), there's also `Swarm.NodeID`, which is unique in the cluster, and is the ID that's used when running `docker node ls` (was even considering to copy that value here, but of course it wouldn't be there if swarm-mode is not active)

tiborvass replied:
@thaJeztah I'm not sure what to do with the swarmkit NodeID. I'd like to unblock this PR though. I personally don't need a UUID check, so I could just check for the non-emptiness of the ID. cc @justincormack

tiborvass commented:
Updated

AkihiroSuda commented:
Probably there are people who are still using an older version of Quay w/o support for schema2

AkihiroSuda commented:
ref: https://github.com/moby/buildkit/issues/409

josephschorr commented:
Hi @tiborvass, TL of Quay here

If manifest V2_1 was deprecated in a PR only 24 days ago, why is the push code being removed so quickly afterwards?

Ignoring the (long but interesting) reasons as to why Quay.io doesn't fully support V2_2 yet (which I'll be discussing in a talk publicly sometime), this fast of a removal seems counter to standard enterprise deprecation guidelines.

Speaking only for myself, I think it would be better to keep V2_1 push for the time being and emit a warning message indicating that an older manifest version is being used and that it will be removed in a future release. In fact, this is what was done when Docker moved from protocol V1 to protocol V2, which allowed users to continue using older versions of registries, but with knowledge that they should upgrade soon.

tiborvass commented:
Hi @josephschorr!

I believe we're on the same page.

A bit of context though: the short-lived v2schema1 was deprecated 3+ years ago in favor of v2schema2 and although we did not have a timeline for complete removal, the intention was clear that v2schema2 deprecates v2schema1. If I'm not mistaken Quay was involved in discussions to migrate to v2schema2 for a long time, and I'm glad to see it happening in the beta program.

When https://github.com/moby/moby/pull/37874 was merged, it not only removed dead code from v1 but also code from v2schema1 which is mainly used by Quay today and would have broken our users who rely on Quay.

We caught that in time so we reverted the v2schema1 removal in https://github.com/moby/moby/pull/39365 to prevent breaking our users, while making sure this time we properly and officially deprecate v2schema1 despite the fact that all players know they should be using v2schema2. That PR does exactly what you describe, adding a warning that v2 schema1 will be removed in a future release: https://github.com/moby/moby/pull/39365/files#diff-8448734ad4bcad711a430d27d303ac16R161

We cherry-picked that into the next release of Docker as well.

What may also cause confusion is that the master branch of moby went back and forth by removing v2schema1, reverting+warning, and reverting the revert. I agree that doing the revert+warning in Docker's release branch would have been simpler. The end goal is for the next release to be the last release with v2schema1 support and have a deprecation warning, but the release after that will have v2schema1 removed. And because that will be based off of moby's current master branch, that is why it is being removed in this PR on moby master.

I strongly hope that Quay's beta program that migrates to v2schema2 is a sign that the timelines for removing v2schema1 will align for both of us.

josephschorr commented:
Hi @tiborvass,

>  although we did not have a timeline for complete removal, the intention was clear that v2schema2 deprecates v2schema1

Definitely, although the lack of communication to end-users presented issues for them to understand why older registry versions no longer work, as the "manifest invalid" error is quite cryptic to those who know about the change. I'm glad this is being fixed now! :)

> What may also cause confusion is that the master branch of moby went back and forth by removing v2schema1, reverting+warning, and this PR that's reverting the revert

Yeah, this is definitely what confused me and I apologize if I came off as rude in my previous message as a result; I was under the (mistaken) impression that this PR was going to once again break users on older installations without warning.

> I strongly hope that Quay's beta program that migrates to v2schema2 is a sign that the timelines for removing v2schema1 will align for both of us.

So, the code is released to our on premises customers and they are using it without issue (AFAIK). Quay.io's delay in full release has, unfortunately, to do with an unrelated issue we've been addressing, but once that issue is fully addressed, we'll be completing the rollout to all our users. I'm *hopeful* this is no more than a few weeks away, but as we have to ensure performance and stability at all costs, its possible we find other blockers as we go along.

On a personal note, I do highly appreciate this change - its making the delays on our side much less stressful, so thank you! :)

josephschorr commented:
Hi @tiborvass,

We have customers that have pulled the code with this warning and are now seeing it when *pulling* manifest V2_1 images, which I believe is a mistake; it was my understanding that the V2_1 code being removed in the future was only for *push*, and that support would remain for pull to support older images that were pushed to registries years ago.

Any ideas if we could get the message changed to only apply on push?

josephschorr commented:
@tiborvass Ping! Any updates on this?

tiborvass commented:
@josephschorr opened https://github.com/moby/moby/pull/39736

josephschorr commented:
@tiborvass Thanks!

dmcgowan reviewed (approved):
LGTM

tonistiigi commented on integration-cli/docker_cli_by_digest_test.go:
```diff
@@ -80,10 +77,6 @@ func (s *DockerRegistrySuite) TestPullByTagDisplaysDigest(c *check.C) {
 	testPullByTagDisplaysDigest(c)
 }
 
-func (s *DockerSchema1RegistrySuite) TestPullByTagDisplaysDigest(c *check.C) {
```
The PR claims to be removing push, so why are these tests being removed?

dmcgowan replied:
The whole schema1 test suite must be removed since that old version of the registry is no longer supported with Docker. If you can't push the image, you can't run the pull test. A new test which pulls an old version from the hub could be added, but that is fragile.

tonistiigi replied:
These tests were added after a CVE fix so it seems weird to remove it without an alternative as the pull code is still remaining(probably for years) and needs to be validated.

andrewhsu commented:
Just chatted with @dmcgowan and the changes in this PR are desired for going into PR #38738

Commit:
Remove push tests and move UUID tests to integration

Partial revert of f23a51a8603a92b124e5037106958baf8c70c5e1.

Only the schema1 push tests are removed but the schema1 pull tests
are still desired.

The UUID test is moved from integration-cli to integration.

Signed-off-by: Tibor Vass <tibor@docker.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

AkihiroSuda commented:
needs rebase

thaJeztah commented:
ping @tiborvass

cpuguy83 commented:
```
[2020-06-25T19:26:01.474Z] can't load test package: package github.com/docker/docker/integration/system (test)
[2020-06-25T19:26:01.474Z] 	imports gotest.tools/assert: cannot find package "gotest.tools/assert" in any of:
[2020-06-25T19:26:01.474Z] 	/go/src/github.com/docker/docker/vendor/gotest.tools/assert (vendor tree)
[2020-06-25T19:26:01.474Z] 	/usr/local/go/src/gotest.tools/assert (from $GOROOT)
[2020-06-25T19:26:01.474Z] 	/go/src/gotest.tools/assert (from $GOPATH)
```

Looks like it needs a gotest.tools/v3 or whatever.

thaJeztah commented:
Did a quick rebase and fixed the gotest.tools import paths to use `v3`

thaJeztah commented:
Seeing a number of failures having `tag invalid`;

```
=== RUN   TestDockerSchema1RegistrySuite/TestPullImageWithAliases
    --- FAIL: TestDockerSchema1RegistrySuite/TestPullImageWithAliases (0.25s)
        cli.go:29: assertion failed: 
            Command:  /usr/local/cli/docker push 127.0.0.1:5000/dockercli/busybox:recent
            ExitCode: 1
            Error:    exit status 1
            Stdout:   The push refers to repository [127.0.0.1:5000/dockercli/busybox]
            1be74353c3d0: Preparing
            1be74353c3d0: Pushed
            
            Stderr:   tag invalid
            
            
            Failures:
            ExitCode was 1 expected 0
            Expected no error
        check_test.go:223: [d152d09ddf8a9] daemon is not started
```

thaJeztah commented:
Also need to look at https://github.com/moby/moby/issues/41095 to see if we can remove that

thaJeztah commented:
Looks like it's because we push to a v1 registry to verify that pulling still works; https://github.com/moby/moby/blob/0fc914de5c9cf13cfa972dbc0ca8034dec7c971d/integration-cli/docker_cli_pull_local_test.go#L38

thaJeztah commented:
Discussing what options we have with @tonistiigi - we could store the registry state in an image and pull that so that we don't have to push to those registries first (or perhaps prepare an image with the registry + data); need to think of options we have there, and what would  work best

thaJeztah commented:
closing this for now; a rebased version of this is in https://github.com/moby/moby/pull/42300, but closed for now; https://github.com/moby/moby/pull/42300#issuecomment-984938579


> Discussing this in the maintainers meeting, and there are still concerns about deprecating this before containerd and others deprecate it, so the request is to instead create a PR to disable it by default, and make support for this opt-in.
>
> Closing this for now
